FreeRTOS-Plus/Demo/FreeRTOS_Plus_PKCS11_Windows_Simulator/examples/objects.c

```diff
@@ -0,0 +1,405 @@
+    "I6Bmxdh4/IpufaGUJ3r7oYal0Q3grV7AA28bXI+Mg+Uzz/6I49Tthtn7VPONnTq4\n" \
+    "aVM+700CgYEA/tntlWjB//UEDahsvyBmgJOUyeZuYG1asbFKG515rncBBHrI3QqW\n" \
+    "78rVVqCQl4w1W0hh3rHmaRC8XneWljGyZ264672wmaoy8T6eOiDiQITfxk3ayhjQ\n" \
+    "Yw1IcCvyUzes6RrbKW3qwIFc0qwhMU0t8iofr8plaGktJs+iTtoqbjMCgYEAwY69\n" \
+    "RcPbSFrl7rAFg34tZss0nJC65QoGO9eEnX1UA4qW3vKEq6WKRNYrjPu4Cb3cfVQb\n" \
+    "4I3pnZDQXUJ8sBX89UGxwlPVEPtQVjyhf9vH3zVTHHgfz1m4P0VD183Bj+mDkQsN\n" \
+    "1LuVMaQiZRU2C+yyflJL2n88fJXNBD2+L09F0PsCgYEAq94COIF2W4w8GI3eEKLk\n" \
+    "HY/phqNtCoCU6kK6zJbNqde9iuzXKyJqGfq7FvmfqWQ304Cj6BAnZ/GOZ1LU7iND\n" \
+    "wjdMZEb/oxHB2ym0gVUN4B773ntAsow7dKFPmb45ZN+mL2oaJP1olDxpgfXtSJdZ\n" \
+    "BfhceCvRabGMbqEWP4EnnjsCgYBgzo1nuT+RfrUFnlADpjX69aNIszJmn47imzYe\n" \
+    "trIwWz5+H3mvh8oSB7QcfUFczfLbVBflRdUUE3vDSR+nyjeMSuEnOlqUdb2GhCia\n" \
+    "FEhxIPCCiqoOQOdf7bqy0pv+6GH/RzIRtgg2ncr+F4/Cgw37FBoQHBcakKnLdz4J\n" \
+    "13IpxQKBgAEj3FDco2hHjdmEVCVXTVA+SPhwkTlzaMk/GkS9ncteaJrHgPaTc3i2\n" \
+    "PJz7uuDt64hf5O9Zi5ecXsPueG8DN3oIPMoFbo/mA2A4ZBUnB8BDrC+3inYuDRxu\n" \
+    "GdWIVVVleMjIDhuhavfzmfBgoir2wQgwnJLPd0vOMDafC6WlcN17\n"             \
+    "-----END RSA PRIVATE KEY-----"
+
+/**
+ * prvObjectImporting covers how to import a private RSA key that was
+ * not generated by the Cryptoki library.
+ *
+ * Note: This method can also be used with certificates and public keys.
+ * This Cryptoki library does not support importing RSA public keys.
+ */
+static void prvObjectImporting( void );
+
+/**
+ * prvObjectGeneration covers how to create a public key and private key pair
+ * with Cryptoki defined attributes using C_GenerateKeyPair.
+ *
+ * Note: The "sign-verify.c" demo has a dependency on the objects created
+ * in this function, and will not work without first running this function.
+ */
+static void prvObjectGeneration( void );
+
+/* Length parameters for importing RSA-2048 private keys. */
+#define MODULUS_LENGTH        pkcs11RSA_2048_MODULUS_BITS / 8
+#define E_LENGTH              3
+#define D_LENGTH              pkcs11RSA_2048_MODULUS_BITS / 8
+#define PRIME_1_LENGTH        128
+#define PRIME_2_LENGTH        128
+#define EXPONENT_1_LENGTH     128
+#define EXPONENT_2_LENGTH     128
+#define COEFFICIENT_LENGTH    128
+
+/**
+ * This function details how to use the PKCS #11 "Object" functions to
+ * manage the objects abstracted by cryptoki.
+ *
+ * http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/os/pkcs11-base-v2.40-os.html
+ * please consult the standard for more information.
+ *
+ * The standard has grouped the functions presented in this demo as:
+ * Object Management Functions.
+ *
+ */
+void vPKCS11ObjectDemo( void )
+{
+    configPRINTF( ( "\r\nStarting PKCS #11 Objects Demo.\r\n" ) );
+
+    /* PKCS #11 defines objects as "An item that is stored on a token. May be
+     * data, a certificate, or a key." This demo will show how to create objects
+     * that are managed by Cryptoki. */
+    prvObjectImporting();
```

Comment:
A dump question... why not just call ```prvObjectImporting()``` and ```prvObjectGeneration()``` at one layer above? Since the two demos seem to be irrelevant to each other.... (Usually with back to back calls, I'm kind of assuming it implies a designed flow. Where as in this case, we can call these two separately or reversed order. )

From my understanding... ```prvObjectImporting()``` parses an existing private key, and call ```C_CreateObject``` to recreate a key based on the information extracted from the existing key. And ```prvObjectGeneration()``` generates key pair. So it sounds like they are two work flows for IoT products.

Reply:
Good point, I will update this

Reply:
Hmm actually,  i'd like to keep it as it is so the demo can scale up if there are more examples to be added for this, for example importing a certificate. This way it keeps the bloat down in the demo header file and main.c

Reply:
(I just noticed you commented back, apologies for the delay. 80%+ of my emails go to trash directly... )

Keeping it like this is certainly fine. In the long run, maybe we want to have white paper equivalent sort of things, to call out different workflows with certificates. Or maybe on demo webpage, some wording like -- "if device is capable, we recommend ...". (All good practices could go here. Like don't load keys into memory, use secure element, ...... ) Otherwise, user could generate keys with IoT Core and load private key into memory etc.

You might have already done so in demo page, which I haven't quite read through. This comment is left in ignorance of relevant existing documentation.